### ISSUE TITLE
DOCS: Simplest example for total newbies

### ISSUE THREAD

**Issue by keyscores** [vsivsi/meteor-job-collection#172]
@vsiverson this package is excellent, solved many problems. Maybe I can contribute to the docs.

I put together a small hello world example as a tutorial for my team. Other people may find it useful. 
https://github.com/keyscores/simplest_example_meteor_job_collection

A few questions came up. We can address in another thread if you prefer.

server.js, line 11 :  When is the right time to start a queue, I would assume on startup, but that seems to fail.

server.js, line 27 : What is callback(), and how can the dev modify it? If s/he can't why are we exposing it?

server.js, line 48 : I'm setting a repeat to 10ms, but sometimes it takes 15 secs to repeat. If I set to 1ms it's instant. What could be blocking it?

**Comment by greatramu** [comment]
Hello! here are the insights, based on my experience with this package

1. jobQueue outside startup seems to work fine for me.
2. "callback()" marks the end of the async function of the worker. You must call this callback to close the job worker so that it can pick up the next available job in the queue. Failing to call this callback will cause the job to run forever and the worker will be lost.
3. I see that you did not configure the jc.promote() and the default promotion time is 15 secs. The job queue starts giving the job to the worker only when it is in "ready" status. And this "promote()" function promotes a job from waiting status to ready.

**Comment by keyscores** [comment]
Thanks! @greatramu 

1. Inside startup it seems to fail. Is there any reason for this?

2. So why can't the package handle this case without the developer input. If it always needs to be called, then the job-collection should handle this. Just seems a bit fragile..

3. Great, can we improve the docs on this? So in my case it is never exactly 15 seconds. And if I set the wait: 1 millisecond, then I get instant tasks. Any reason for this?

**Comment by vsivsi** [comment]
@greatramu Thanks for contributing your answers!  They look good.

@keyscores 

FYI, There are two working sample applications for job-collection:
https://github.com/vsivsi/meteor-job-collection-playground
https://github.com/vsivsi/meteor-file-job-sample-app

Answers to your new questions:

   1. No, it shouldn't.
   2. See: https://github.com/vsivsi/meteor-job-collection/issues/167
   3. See: https://github.com/vsivsi/meteor-job-collection#jcpromotemilliseconds---server-only  I'd be happy to accept a PR for that section that makes it clearer. Meteor and node.js are a single threaded async system. Everything is "best effort", there are no timing guarantees in this system. I don't recommend setting `jc.promote()` to 1ms, because that means you will be querying the underlying MongoDB collection 1000x per second just to try to promote waiting jobs more quickly! Probably a very bad idea!

**Comment by keyscores** [comment]
@vsivsi thanks for the feedback. 

I have seen your sample apps and they were very helpful! I needed something a bit more hello world, in JS,  and maybe other users do as well.

I'll do a PR on the docs as soon as I can.
Thanks again!

**Comment by keyscores** [comment]
@vsivsi I'm working on a Docs PR.

One think I want to make clear is how to run a Meteor method within a task, and have the status update correctly

One issue I have in production, is that I don't duplicate jobs in queue. ,If an identical one is already in ready/waiting mode, I don't add it. However I get a false negative if I use a Meteor Method. The job.done() gets called immediately before the return.

I'm guessing we must use futures to prevent the job.done() from being called prematurely. Can you check this implementation? Is there a better way?
https://github.com/keyscores/simplest_example_meteor_job_collection/blob/master/server/longWait.js

**Comment by vsivsi** [comment]
Right, most any worker function that does interesting work will need to invoke asynchronous functions (Meteor method calls, DB accesses, file I/O, HTTP requests...) So the worker function needs to be written to only invoke `job.done()` once the job is *really* done. This is the essence of async programming. There are lots of tools in the node.js universe for managing this complexity (async.js, Promises and Fibers, just being a few). 

In your "hello world" scenario, putting the `job.done()` in the method callback is entirely appropriate, however, the `callback()` call for the worker function should also be within that method callback. It should never be invoked until either `job.done()` or `job.fail()` are called! Otherwise more jobs will start on that worker before the previous one is completed!

**Comment by juliomac** [comment]
Hi @keyscores, thanks for your newbie efforts! Being a newbie myself, I am Looking forward to it! I quite couldn't manage to understand the examples in your repo yet, however.

What I am trying to accomplish is to have a job created as soon as a user is enrolled (accounts package). The job then checks if user created a password (i.e. read the enrollment e-mail, clicked on the link and created a password) and, if if hasn't for the last 3 days, then resend a new e-mail with a renewed enrollment link.

How could I do this with this package? Is it appropriate for the task?

**Comment by greatramu** [comment]
Hi juliomac,  I think this package is overkill for your purpose. Long running jobs just for sending emails does not sound nice. In my view, this package adds value when you want to offload a chunk of processing (heavy) work from the main process - like image processing for example.

For your task, it is much simpler to have a cron job (there is a meteor package for synched cron) running which monitors your users collection every night and sends out emails when necessary.

**Comment by keyscores** [comment]
@juliomac You could use vsivsi's  Job Collection, to set up recurring job as @greatramu suggests. See using jc.later() in the docs. I'm not sure it's the best thing, since Job Collection adds another level of complexity. Try some other packages, but if you really want Job Collection, try  something like: 
```  
    /// START THE SERVER
   ///CLEANQUEUE is new job queue

    CLEANQUEUE.startJobServer();
    CLEANQUEUE.processJobs(
      // Type of job to request
      ['clean'], null , //no options
      function (job, callback) {
        // ##### THIS IS WHERE THE WORK GETS DONE #####//
        Meteor.call("pipeline", function(){
          job.done();
          callback(); // Mark the job done
        });
      });

//// EACH JOB
    importJob = new Job( CLEANQUEUE , 'clean', {} );

    importJob.repeat({
      // UTC TIME !
      schedule: CLEANQUEUE.later.parse.text('at 4:00 am') //CLEANQUEUE.later.parse.text('every 10 seconds') //
    })

    importJob.save()
```

**Comment by juliomac** [comment]
Thanks @greatramu  for the advice! I though this package was a Cron job scheduler.

**Comment by juliomac** [comment]
Thanks for the nice example @keyscores ! From what you and @greatramu said, it really seems an overkill for my use case. I will go with [percolatestudio/meteor-synced-cron](https://github.com/percolatestudio/meteor-synced-cron) . It seems to be highly popular in Atmosphere.
